### ISSUE TITLE
An error occurred: E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.

### ISSUE THREAD

**Issue by konomikitten** [mvo5/unattended-upgrades#292]
Unattended upgrades is currently spitting this error for me:

```
$ sudo unattended-upgrade --dry-run --verbose
Checking if system is running on battery is skipped. Please install powermgmt-base package to check power status and skip installing updates when the system is running on battery.
Initial blacklist :
Initial whitelist:
Starting unattended upgrades script
Allowed origins are: o=Debian,n=buster, o=Debian,n=buster-updates, o=Debian Backports,n=buster-backports, o=Unofficial Multimedia Packages,n=buster
package ure upgradable but fails to be marked for upgrade (E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.)
package xterm upgradable but fails to be marked for upgrade (E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.)
An error occurred: E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.
Traceback (most recent call last):
  File "/usr/bin/unattended-upgrade", line 1663, in main
    install_start_time)
  File "/usr/bin/unattended-upgrade", line 1852, in run
    mark_pkgs_to_upgrade(cache, pkgs_to_upgrade)
  File "/usr/bin/unattended-upgrade", line 1702, in mark_pkgs_to_upgrade
    from_user=not pkg.is_auto_installed)
  File "/usr/bin/unattended-upgrade", line 237, in mark_upgrade_adjusted
    self.call_adjusted(apt.package.Package.mark_upgrade, pkg, **kwargs)
  File "/usr/bin/unattended-upgrade", line 207, in call_adjusted
    function(pkg, **kwargs)
  File "/usr/lib/python3/dist-packages/apt/package.py", line 1564, in mark_upgrade
    self.mark_install(from_user=from_user)
  File "/usr/lib/python3/dist-packages/apt/package.py", line 1556, in mark_install
    fixer.resolve(True)
apt_pkg.Error: E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.
Traceback (most recent call last):
  File "/usr/bin/unattended-upgrade", line 2195, in <module>
    sys.exit(main(options))
  File "/usr/bin/unattended-upgrade", line 1663, in main
    install_start_time)
  File "/usr/bin/unattended-upgrade", line 1852, in run
    mark_pkgs_to_upgrade(cache, pkgs_to_upgrade)
  File "/usr/bin/unattended-upgrade", line 1702, in mark_pkgs_to_upgrade
    from_user=not pkg.is_auto_installed)
  File "/usr/bin/unattended-upgrade", line 237, in mark_upgrade_adjusted
    self.call_adjusted(apt.package.Package.mark_upgrade, pkg, **kwargs)
  File "/usr/bin/unattended-upgrade", line 207, in call_adjusted
    function(pkg, **kwargs)
  File "/usr/lib/python3/dist-packages/apt/package.py", line 1564, in mark_upgrade
    self.mark_install(from_user=from_user)
  File "/usr/lib/python3/dist-packages/apt/package.py", line 1556, in mark_install
    fixer.resolve(True)
apt_pkg.Error: E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.
bash ~
$
```

Someone from the Debian irc channel suggested I hold the mentioned packages I did `sudo aptitude hold ure` and then unattended upgrades completes without bombing to an exception:

```
$ sudo unattended-upgrade --dry-run --verbose
Checking if system is running on battery is skipped. Please install powermgmt-base package to check power status and skip installing updates when the system is running on battery.
Initial blacklist :
Initial whitelist:
Starting unattended upgrades script
Allowed origins are: o=Debian,n=buster, o=Debian,n=buster-updates, o=Debian Backports,n=buster-backports, o=Unofficial Multimedia Packages,n=buster
package libreoffice-core upgradable but fails to be marked for upgrade (E:Unable to correct problems, you have held broken packages.)
package libreoffice-gtk2 upgradable but fails to be marked for upgrade (E:Unable to correct problems, you have held broken packages.)
package uno-libs3 upgradable but fails to be marked for upgrade (E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.)
Option --dry-run given, *not* performing real actions
Packages that will be upgraded: avahi-daemon base-files debian-archive-keyring exim4 exim4-base exim4-config exim4-daemon-light fonts-opensymbol gir1.2-javascriptcoregtk-4.0 gir1.2-webkit2-4.0 groff-base iputils-ping libavahi-client3 libavahi-common-data libavahi-common3 libavahi-compat-libdnssd1 libavahi-core7 libavahi-glib1 libbsd0 libcpupower1 libjavascriptcoregtk-4.0-18 libjs-jquery liblirc-client0 libnss-systemd libpam-systemd libpq5 libpython3.7 libpython3.7-dev libpython3.7-minimal libpython3.7-stdlib libreoffice-common libreoffice-style-colibre libreoffice-style-tango libsystemd0 libudev1 libwebkit2gtk-4.0-37 linux-cpupower linux-image-amd64 linux-libc-dev python3.7 python3.7-dev python3.7-minimal systemd systemd-sysv udev xterm
Writing dpkg log to /var/log/unattended-upgrades/unattended-upgrades-dpkg.log
apt-listchanges: Reading changelogs...
Extracting templates from packages: 100%
Preconfiguring packages ...
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --unpack --auto-deconfigure /var/cache/apt/archives/base-files_10.3+deb10u9_amd64.deb
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --configure base-files:amd64
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --unpack --auto-deconfigure /var/cache/apt/archives/libsystemd0_241-7~deb10u7_amd64.deb
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --configure libsystemd0:amd64
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --unpack --auto-deconfigure /var/cache/apt/archives/libnss-systemd_241-7~deb10u7_amd64.deb /var/cache/apt/archives/libpam-systemd_241-7~deb10u7_amd64.deb /var/cache/apt/archives/systemd_241-7~deb10u7_amd64.deb /var/cache/apt/archives/udev_241-7~deb10u7_amd64.deb /var/cache/apt/archives/libudev1_241-7~deb10u7_amd64.deb
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --configure libudev1:amd64 systemd:amd64
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --unpack --auto-deconfigure --recursive /tmp/apt-dpkg-install-fWdyFu
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --configure debian-archive-keyring:all
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --unpack --auto-deconfigure --recursive /tmp/apt-dpkg-install-ycAhQH
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --configure --pending
All upgrades installed
bash ~
$
```

Here is a detail list of what upgrades the system wants to perform:
```
$ aptitude -F '%c%a%M %p %t %v %V' search ~U
i A avahi-daemon                                    stable                       0.7-4+b1       0.7-4+deb10u1
i   base-files                                      stable                       10.3+deb10u8   10.3+deb10u9
i   debian-archive-keyring                          stable                       2019.1         2019.1+deb10u1
i   exim4                                           stable                       4.92-8+deb10u4 4.92-8+deb10u5
i   exim4-base                                      stable                       4.92-8+deb10u4 4.92-8+deb10u5
i   exim4-config                                    stable                       4.92-8+deb10u4 4.92-8+deb10u5
i   exim4-daemon-light                              stable                       4.92-8+deb10u4 4.92-8+deb10u5
i   fonts-opensymbol                                stable                       2:102.10+LibO6 2:102.10+LibO6
i A gir1.2-javascriptcoregtk-4.0                    stable                       2.30.5-1~deb10 2.30.6-1~deb10
i A gir1.2-webkit2-4.0                              stable                       2.30.5-1~deb10 2.30.6-1~deb10
i   groff-base                                      stable                       1.22.4-3       1.22.4-3+deb10
i   iputils-ping                                    stable                       3:20180629-2+d 3:20180629-2+d
i A libavahi-client3                                stable                       0.7-4+b1       0.7-4+deb10u1
i A libavahi-common-data                            stable                       0.7-4+b1       0.7-4+deb10u1
i A libavahi-common3                                stable                       0.7-4+b1       0.7-4+deb10u1
i A libavahi-compat-libdnssd1                       stable                       0.7-4+b1       0.7-4+deb10u1
i A libavahi-core7                                  stable                       0.7-4+b1       0.7-4+deb10u1
i   libavahi-glib1                                  stable                       0.7-4+b1       0.7-4+deb10u1
i A libbsd0                                         stable                       0.9.1-2        0.9.1-2+deb10u
i A libcpupower1                                    stable                       4.19.171-2     4.19.181-1
i A libjavascriptcoregtk-4.0-18                     stable                       2.30.5-1~deb10 2.30.6-1~deb10
i A libjs-jquery                                    stable                       3.3.1~dfsg-3   3.3.1~dfsg-3+d
i A liblirc-client0                                 stable                       0.10.1-6.2~deb 0.10.1-6.3~deb
i A libnss-systemd                                  stable                       241-7~deb10u6  241-7~deb10u7
i   libpam-systemd                                  stable                       241-7~deb10u6  241-7~deb10u7
i A libpq5                                          stable                       11.10-0+deb10u 11.11-0+deb10u
i A libpython3.7                                    stable                       3.7.3-2+deb10u 3.7.3-2+deb10u
i A libpython3.7-dev                                stable                       3.7.3-2+deb10u 3.7.3-2+deb10u
i A libpython3.7-minimal                            stable                       3.7.3-2+deb10u 3.7.3-2+deb10u
i A libpython3.7-stdlib                             stable                       3.7.3-2+deb10u 3.7.3-2+deb10
i   libreoffice-common                              stable                       1:6.1.5-3+deb1 1:6.1.5-3+deb1
i A libreoffice-core                                stable                       1:6.1.5-3+deb1 1:6.1.5-3+deb1
i A libreoffice-gtk2                                stable                       1:6.1.5-3+deb1 1:6.1.5-3+deb1
i A libreoffice-style-colibre                       stable                       1:6.1.5-3+deb1 1:6.1.5-3+deb1
i A libreoffice-style-tango                         stable                       1:6.1.5-3+deb1 1:6.1.5-3+deb1
i   libsystemd0                                     stable                       241-7~deb10u6  241-7~deb10u7
i   libudev1                                        stable                       241-7~deb10u6  241-7~deb10u7
i A libwebkit2gtk-4.0-37                            stable                       2.30.5-1~deb10 2.30.6-1~deb10
i   linux-cpupower                                  stable                       4.19.171-2     4.19.181-1
i   linux-image-amd64                               stable                       4.19+105+deb10 4.19+105+deb10
i A linux-libc-dev                                  stable                       4.19.171-2     4.19.181-1
i A python3.7                                       stable                       3.7.3-2+deb10u 3.7.3-2+deb10u
i A python3.7-dev                                   stable                       3.7.3-2+deb10u 3.7.3-2+deb10u
i A python3.7-minimal                               stable                       3.7.3-2+deb10u 3.7.3-2+deb10u
i   systemd                                         stable                       241-7~deb10u6  241-7~deb10u7
i   systemd-sysv                                    stable                       241-7~deb10u6  241-7~deb10u7
i   udev                                            stable                       241-7~deb10u6  241-7~deb10u7
i A uno-libs3                                       stable                       6.1.5-3+deb10u 6.1.5-3+deb10u
i   ure                                             stable                       6.1.5-3+deb10u 6.1.5-3+deb10u
i A xterm                                           stable                       344-1          344-1+deb10u1
bash ~
$ 
```

**Comment by konomikitten** [comment]
I ended up upgrading the packages that unattended-upgrades was error out on manually with aptitude as I did not want to wait to update the system, which seemed to resolve the issue. **I still think unattended-upgrades should not be erroring out like this**.

```
$ sudo unattended-upgrade --dry-run --verbose
Checking if system is running on battery is skipped. Please install powermgmt-base package to check power status and skip installing updates when the system is running on battery.
Initial blacklist :
Initial whitelist:
Starting unattended upgrades script
Allowed origins are: o=Debian,n=buster, o=Debian,n=buster-updates, o=Debian Backports,n=buster-backports, o=Unofficial Multimedia Packages,n=buster
package ure upgradable but fails to be marked for upgrade (E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.)
package xterm upgradable but fails to be marked for upgrade (E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.)
An error occurred: E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.
Traceback (most recent call last):
  File "/usr/bin/unattended-upgrade", line 1663, in main
    install_start_time)
  File "/usr/bin/unattended-upgrade", line 1852, in run
    mark_pkgs_to_upgrade(cache, pkgs_to_upgrade)
  File "/usr/bin/unattended-upgrade", line 1702, in mark_pkgs_to_upgrade
    from_user=not pkg.is_auto_installed)
  File "/usr/bin/unattended-upgrade", line 237, in mark_upgrade_adjusted
    self.call_adjusted(apt.package.Package.mark_upgrade, pkg, **kwargs)
  File "/usr/bin/unattended-upgrade", line 207, in call_adjusted
    function(pkg, **kwargs)
  File "/usr/lib/python3/dist-packages/apt/package.py", line 1564, in mark_upgrade
    self.mark_install(from_user=from_user)
  File "/usr/lib/python3/dist-packages/apt/package.py", line 1556, in mark_install
    fixer.resolve(True)
apt_pkg.Error: E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.
Traceback (most recent call last):
  File "/usr/bin/unattended-upgrade", line 2195, in <module>
    sys.exit(main(options))
  File "/usr/bin/unattended-upgrade", line 1663, in main
    install_start_time)
  File "/usr/bin/unattended-upgrade", line 1852, in run
    mark_pkgs_to_upgrade(cache, pkgs_to_upgrade)
  File "/usr/bin/unattended-upgrade", line 1702, in mark_pkgs_to_upgrade
    from_user=not pkg.is_auto_installed)
  File "/usr/bin/unattended-upgrade", line 237, in mark_upgrade_adjusted
    self.call_adjusted(apt.package.Package.mark_upgrade, pkg, **kwargs)
  File "/usr/bin/unattended-upgrade", line 207, in call_adjusted
    function(pkg, **kwargs)
  File "/usr/lib/python3/dist-packages/apt/package.py", line 1564, in mark_upgrade
    self.mark_install(from_user=from_user)
  File "/usr/lib/python3/dist-packages/apt/package.py", line 1556, in mark_install
    fixer.resolve(True)
apt_pkg.Error: E:Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.
bash ~
$
```

```
bash ~
$ sudo aptitude safe-upgrade ure xterm
Resolving dependencies...
The following packages will be upgraded:
  uno-libs3 ure xterm
3 packages upgraded, 0 newly installed, 0 to remove and 47 not upgraded.
Need to get 2,645 kB/3,416 kB of archives. After unpacking 0 B will be used.
Do you want to continue? [Y/n/?] y
Get: 1 https://deb.debian.org/debian buster/main amd64 ure amd64 6.1.5-3+deb10u7 [1,747 kB]
Get: 2 https://deb.debian.org/debian buster/main amd64 uno-libs3 amd64 6.1.5-3+deb10u7 [898 kB]
Fetched 2,645 kB in 8s (317 kB/s)
Reading changelogs... Done
(Reading database ... 126335 files and directories currently installed.)
Preparing to unpack .../ure_6.1.5-3+deb10u7_amd64.deb ...
Unpacking ure (6.1.5-3+deb10u7) over (6.1.5-3+deb10u6) ...
Preparing to unpack .../uno-libs3_6.1.5-3+deb10u7_amd64.deb ...
Unpacking uno-libs3 (6.1.5-3+deb10u7) over (6.1.5-3+deb10u6) ...
Preparing to unpack .../xterm_344-1+deb10u1_amd64.deb ...
Unpacking xterm (344-1+deb10u1) over (344-1) ...
Setting up xterm (344-1+deb10u1) ...
Setting up uno-libs3 (6.1.5-3+deb10u7) ...
Setting up ure (6.1.5-3+deb10u7) ...
Processing triggers for mime-support (3.62) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
Processing triggers for libc-bin (2.28-10) ...
Processing triggers for man-db (2.8.5-2) ...
Processing triggers for libreoffice-common (1:6.1.5-3+deb10u6) ...
Processing triggers for desktop-file-utils (0.23-4) ...

Current status: 47 (-3) upgradable.
bash ~
$
```

```
bash ~
$ sudo unattended-upgrade --dry-run --verbose
Checking if system is running on battery is skipped. Please install powermgmt-base package to check power status and skip installing updates when the system is running on battery.
Initial blacklist :
Initial whitelist:
Starting unattended upgrades script
Allowed origins are: o=Debian,n=buster, o=Debian,n=buster-updates, o=Debian Backports,n=buster-backports, o=Unofficial Multimedia Packages,n=buster
Option --dry-run given, *not* performing real actions
Packages that will be upgraded: avahi-daemon base-files debian-archive-keyring exim4 exim4-base exim4-config exim4-daemon-light fonts-opensymbol gir1.2-javascriptcoregtk-4.0 gir1.2-webkit2-4.0 groff-base iputils-ping libavahi-client3 libavahi-common-data libavahi-common3 libavahi-compat-libdnssd1 libavahi-core7 libavahi-glib1 libbsd0 libcpupower1 libjavascriptcoregtk-4.0-18 libjs-jquery liblirc-client0 libnss-systemd libpam-systemd libpq5 libpython3.7 libpython3.7-dev libpython3.7-minimal libpython3.7-stdlib libreoffice-common libreoffice-core libreoffice-gtk2 libreoffice-style-colibre libreoffice-style-tango libsystemd0 libudev1 libwebkit2gtk-4.0-37 linux-cpupower linux-image-amd64 linux-libc-dev python3.7 python3.7-dev python3.7-minimal systemd systemd-sysv udev
Writing dpkg log to /var/log/unattended-upgrades/unattended-upgrades-dpkg.log
apt-listchanges: Reading changelogs...
Extracting templates from packages: 100%
Preconfiguring packages ...
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --unpack --auto-deconfigure /var/cache/apt/archives/base-files_10.3+deb10u9_amd64.deb
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --configure base-files:amd64
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --unpack --auto-deconfigure /var/cache/apt/archives/libsystemd0_241-7~deb10u7_amd64.deb
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --configure libsystemd0:amd64
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --unpack --auto-deconfigure /var/cache/apt/archives/libnss-systemd_241-7~deb10u7_amd64.deb /var/cache/apt/archives/libpam-systemd_241-7~deb10u7_amd64.deb /var/cache/apt/archives/systemd_241-7~deb10u7_amd64.deb /var/cache/apt/archives/udev_241-7~deb10u7_amd64.deb /var/cache/apt/archives/libudev1_241-7~deb10u7_amd64.deb
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --configure libudev1:amd64 systemd:amd64
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --unpack --auto-deconfigure --recursive /tmp/apt-dpkg-install-Ib6qg5
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --configure debian-archive-keyring:all
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --no-triggers --unpack --auto-deconfigure --recursive /tmp/apt-dpkg-install-zBGkSk
/usr/bin/dpkg --force-confdef --force-confold --status-fd 11 --configure --pending
All upgrades installed
bash ~
$
```

**Comment by rbalint** [comment]
This is fixed in 1.11.2 and I while this traceback is ugly it is sent in the email report. You can verify that by installing 1.11.2 from 
http://snapshot.debian.org/package/unattended-upgrades/1.11.2/ .

I've tried to backport the numerous fixes that went in for 1.11.2, but that did not go through:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=928108

**Comment by konomikitten** [comment]
> I've tried to backport the numerous fixes that went in for 1.11.2, but that did not go through

That's frustrating, thanks for all your hard work since this is fixed in 1.11.2 I'm closing this report.

**Comment by konomikitten** [comment]
Hmm did I close this a little too quickly? I already had version 1.11.2 installed according to apt policy:
```
$ apt policy unattended-upgrades
unattended-upgrades:
  Installed: 1.11.2
  Candidate: 1.11.2
  Version table:
 *** 1.11.2 500
        500 https://deb.debian.org/debian buster/main amd64 Packages
        500 https://deb.debian.org/debian buster/main i386 Packages
        100 /var/lib/dpkg/status
```

**Comment by rbalint** [comment]
I'm sorry, I meant it is fixed in 1.12:
http://snapshot.debian.org/package/unattended-upgrades/1.12/
in this commit:
https://github.com/mvo5/unattended-upgrades/commit/b864d2837b765efd173aef41f9b704842744a671

**Comment by konomikitten** [comment]
All fine keeping this closed then thanks for you time.
